test/common/lcov.py

```diff
@@ -358,10 +371,98 @@ def covranges(functions):
+                       "line": cur_line}
+            if start_line != cur_line:
+                comment["start_line"] = start_line
+                body = f"These {cur_line - start_line + 1} added lines are not executed by any test."
+            else:
+                body = "This added line is not executed by any test."
+            if ta_url:
+                body += f"  [Details]({ta_url}/Coverage/lcov/github-pr.diff.gcov.html)"
+            comment["body"] = body
+            comments.append(comment)
+
+    dm = DiffMap("lcov/github-pr.diff")
+
+    with open(diff_info_file) as f:
+        for line in f.readlines():
+            if line.startswith("DA:"):
+                parts = line[3:].split(",")
+                if int(parts[1]) == 0:
+                    (src, line) = dm.find_source(int(parts[0]))
+                    if src == cur_src and line == cur_line + 1:
+                        cur_line = line
+                    else:
+                        flush_cur_comment()
+                        cur_src = src
+                        start_line = line
+                        cur_line = line
+        flush_cur_comment()
+
+    return comments
+
+def prepare_for_code_coverage():
+    # This gives us a convenient link at the top of the logs, see link-patterns.json
+    print("Code coverage report in Coverage/index.html")
+    try:
+        os.makedirs("lcov", exist_ok=True)
+        with open("lcov/github-pr.diff", "w") as f:
+            subprocess.check_call(["git", "-c", "diff.noprefix=false", "diff", "--patience", "main"], stdout=f)
+    except subprocess.CalledProcessError:
+        pass
+
+def create_coverage_report():
+    output = os.environ.get("TEST_ATTACHMENTS", BASE_DIR)
+    lcov_files = glob.glob(f"{BASE_DIR}/lcov/*.info.gz")
+    try:
+        title = os.path.basename(subprocess.check_output(["git", "remote", "get-url", "origin"]))
+    except subprocess.CalledProcessError:
+        title = "?"
+    if len(lcov_files) > 0:
+        try:
+            all_file = f"{BASE_DIR}/lcov/all.info"
+            diff_file = f"{BASE_DIR}/lcov/diff.info"
+            subprocess.check_call(["lcov", "--output", all_file] +
+                                  sum(map(lambda f: ["--add", f], lcov_files), []))
+            subprocess.check_call(["lcov", "--output", diff_file,
+                                   "--extract", all_file, "*/github-pr.diff"])
+            subprocess.check_call(["genhtml", "--no-function-coverage",
+                                   "--prefix", os.getcwd(),
+                                   "--title", title,
+                                   "--output-dir", f"{output}/Coverage", all_file])
+            comments = get_review_comments(diff_file)
+            rev = os.environ.get("TEST_REVISION", None)
+            pull = os.environ.get("TEST_PULL", None)
+            if rev and pull:
+                api = github.GitHub()
+                old_comments = api.get(f"pulls/{pull}/comments") or []
```

Comment:
From the docs, I've found out, this is limited to 30 comments, so one has to fetch more pages using `page` or setting `per_page` to 100.

https://docs.github.com/en/rest/pulls/comments#list-review-comments-on-a-pull-request

Reply:
100 is probably good enough for now yes :+1: